### ISSUE TITLE
Warn TODOs without comment

### ISSUE THREAD

**Issue by ypresto** [eslint/eslint#6947]
**When does this rule warn? Please describe and show example code:**

```js
// Good
// TODO: I'd like to find a way to make this faster
doSomething();

// Bad
// TODO
doSomething();
```

**Is this rule preventing an error or is it stylistic?**

stylistic

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

It is specified in official code conventions:

> Comments may be used to annotate pieces of code **with additional information**.
> TODO - indicates that the code is not yet complete. **Information about the next steps should be included.**

http://eslint.org/docs/developer-guide/code-conventions#comment-annotations

Also implemented by rubocop (linter for ruby).
https://github.com/bbatsov/rubocop/blob/49598a035cd39a75ec91466a4e1a79c7e4f0f5f9/spec/rubocop/cop/style/comment_annotation_spec.rb

**Are you willing to create the rule yourself?**

yes..!

**Comment by platinumazure** [comment]
I like the idea for sure! One question I have is whether it might fit better as an enhancement for rule `no-warning-comments`. Maybe as an option of warning comments that are only tolerated if they have extra content in the comments... Thoughts on that?

**Comment by ypresto** [comment]
They might be different rules, because one is to disallow and another is to style TODOs.
For example, some user might use `no-warning-comments` only for FIXME (e.g. better to fix now than write it :). In this case user still would use new `no-warnings-without-comment` rules for TODO or HACK.

**Comment by ypresto** [comment]
```js
"no-warning-comments": [<enabled>, {
  "terms": "FIXME",
  "location": "anywhere",
  "terms_without_comment": "TODO", "HACK" // hmm.. Is location affect or not..?
}]
```

**Comment by vitorbal** [comment]
I like this idea as well. What if we would have an option to `no-warning-comments` called `allowIfContainsDescription`:

```json
"no-warning-comments": ["error", {
  "allowIfContainsDescription": "todo"
}]
```

Valid with above config:
```javascript
// TODO handle server errors
```

Invalid with above config:
```javascript
// TODO
```

```javascript
// FIXME this should be made faster
```

**Comment by ypresto** [comment]
That option name sounds great..!
To clarify, are words specified `allowIfContainsDescription` in but not in `terms` always allowed?

```js
"no-warning-comments": [<enabled>, {
  "terms": [],
  "allowIfContainsDescription": ["TODO", "HACK"]
}]
```

```
Valid:

// TODO
// HACK
```

**Comment by vitorbal** [comment]
Good question. I would say yes, if `terms` is empty then `allowIfContainsDescription` doesn't do anything because all terms are being allowed anyway.

 I would be curious what the rest of the team things of this though. I also think the name of the option can be improved.

**Comment by platinumazure** [comment]
I think maybe it should be a separate rule after all. Yes, I know I
advocated the opposite earlier but I've come around.

On Aug 24, 2016 8:52 AM, "Vitor Balocco" <notifications@github.com> wrote:

Good question. I would say yes, if terms is empty then
allowIfContainsDescription doesn't do anything because all terms are being
allowed anyway.

I would be curious what the rest of the team things of this though. I also
think the name of the option can be improved.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/eslint/eslint/issues/6947#issuecomment-242071614>, or mute
the thread

**Comment by vitorbal** [comment]
I would not be against making this another rule as well!

**Comment by nzakas** [comment]
This would have to be a new rule, if anything, as it's a very different use case than `no-warning-comments`. I'm not sure it meets the bar of "ESLint would be incomplete without it," though. http://eslint.org/docs/developer-guide/contributing/new-rules#accepting-a-rule

**Comment by platinumazure** [comment]
Switching to "feature request" tag.

I'm with @nzakas on inclusion in core.

@ypresto Since you're willing to write the rule yourself, this means your best bet is to create either a standalone custom rule or a plugin containing this rule (or, if you find a suitable plugin in the wild, adding this rule to that plugin). We are always happy to help in [our chatroom](https://gitter.im/eslint/eslint) if you need help with creating the rule or plugin. If you want the rule in core, it seems you'll have to convince a few skeptics that ESLint core itself is incomplete without this rule.

**Comment by ypresto** [comment]
> ESLint core itself is incomplete without this rule

It seems almost all items in [Code Conventions](http://eslint.org/docs/developer-guide/code-conventions) have matching rules.
It looks intended to cover all items in conventions by built-in rules, isn't it?

Examples:

- File Format: (no matching rule)
- Indentation: `indent`
- Primitive Literals:
  - Double quotes: `quotes`
  - No wrapping (`"hello \` + linebreak + `world"`): `no-multi-str`
  - Decimal point: no-floating-decimal
  - Octal is deprecated: `no-octal`
  - Disallow `foo != null` when foo may be undefined: `no-eq-null`
  - Disallow undefined: `no-undefined`
- Operator Spacing: `space-infix-ops`
- Parentheses Spacing: `space-in-parens`
- Object Literals
  - Closing brace position: `object-curly-newline`
  - Indentation: `indent`
  - Property line breaking: `object-property-newline`
  - Property quotes: `quote-props`
  - Property colon spacing: `key-spacing`
  - Opening brace positon, wrapping function: (no matching rule)
- etc...

**Comment by ypresto** [comment]
Note that Airbnb style guide has [similar rules](https://github.com/airbnb/javascript#comments--actionitems) for TODO and FIXME.

**Comment by nzakas** [comment]
@ypresto I'm sorry, there's not enough support on the team for this. If you still want the rule, you can always create a [custom rule](http://eslint.org/docs/developer-guide/working-with-rules) instead.

**Comment by ypresto** [comment]
No problem, thanks for considering this rule..!
